src/checkers/gitchecker.py

```diff
@@ -119,12 +122,23 @@ async def _check_has_new(external_data: ExternalGitRepo):
         else:
             sorted_tags = matching_tags
 
-        try:
-            latest_tag = sorted_tags[-1]
-        except IndexError as err:
-            raise CheckerQueryError(
-                f"{external_data.current_version.url} has no tags matching '{tag_pattern}'"
-            ) from err
+        if constraints:
+            matching_versions = [tag.version for tag in matching_tags]
+            versions = filter_versions(matching_versions, constraints, sort=False)
+            try:
+                latest_version = versions[-1]
+                for found_tag in matching_tags:
+                    if found_tag.version is latest_version:
+                        latest_tag = found_tag
```

Comment:
`filter_versions()` supports `to_string` args. If you pass a callable to it, you can filter `TagWithVersion` objects directly, without converting them to version string list first.
E.g. `filter_versions(matching_tags, to_string = lambda t: t.version)` should work.
This way you won't need to iterate over `matching_tags` again, looking for a tag with matching version.

Reply:
Overall, I'd change it like this:
```patch
diff --git a/src/checkers/gitchecker.py b/src/checkers/gitchecker.py
index f4d5bea..3d5db67 100644
--- a/src/checkers/gitchecker.py
+++ b/src/checkers/gitchecker.py
@@ -98,6 +98,7 @@ class GitChecker(Checker):
 
         version_scheme = external_data.checker_data.get("version-scheme", "loose")
         tag_cls = TAG_VERSION_SCHEMES[version_scheme]
+        sort_tags = external_data.checker_data.get("sort-tags", True)
         constraints = external_data.checker_data.get("versions", {}).items()
 
         matching_tags = []
@@ -117,28 +118,24 @@ class GitChecker(Checker):
             version = tag_match.group(1)
             matching_tags.append(tag_cls(commit, tag, annotated, version))
 
-        if external_data.checker_data.get("sort-tags", True):
+        if constraints:
+            sorted_tags = filter_versions(
+                matching_tags,
+                constraints=constraints,
+                to_string=lambda t: t.version,
+                sort=sort_tags,
+            )
+        elif sort_tags:
             sorted_tags = sorted(matching_tags)
         else:
             sorted_tags = matching_tags
 
-        if constraints:
-            matching_versions = [tag.version for tag in matching_tags]
-            versions = filter_versions(matching_versions, constraints, sort=False)
-            try:
-                latest_version = versions[-1]
-                for found_tag in matching_tags:
-                    if found_tag.version is latest_version:
-                        latest_tag = found_tag
-            except IndexError as err:
-                raise CheckerQueryError("Can't find matching version") from err
-        else:
-            try:
-                latest_tag = sorted_tags[-1]
-            except IndexError as err:
-                raise CheckerQueryError(
-                    f"{external_data.current_version.url} has no tags matching '{tag_pattern}'"
-                ) from err
+        try:
+            latest_tag = sorted_tags[-1]
+        except IndexError as err:
+            raise CheckerQueryError(
+                f"{external_data.current_version.url} has no tags matching '{tag_pattern}'"
+            ) from err
 
         new_version = ExternalGitRef(
             url=external_data.current_version.url,

```

Reply:
Thanks, that is much cleaner.